### PR TITLE
Add guard against null-valued settings

### PR DESCRIPTION
**Pull request by jasontedor**
This commit adds a guard against null-valued settings that are loaded
from yaml or json settings files, and also adds a test that ensures
the same remains true for settings loaded from properties files.

Relates #17292

### CODE REVIEW

**Comment by jasontedor** [comment]
This gives more user-friendly error messages like:

```
Exception in thread "main" SettingsException[Failed to load settings from [elasticsearch.yml]]; nested: ElasticsearchParseException[null-valued setting found for key [cluster.name] found at line number [1], column number [14]];
Likely root cause: ElasticsearchParseException[null-valued setting found for key [cluster.name] found at line number [1], column number [14]]
```

instead of

```
Exception in thread "main" java.lang.NullPointerException: value can not be null for [cluster.name]
```

**Comment by rjernst** [comment]
LGTM. I would just call the boolean "allowNullValues" though.

File: core/src/main/java/org/elasticsearch/common/settings/loader/JsonSettingsLoader.java [hunk 1]
```diff
@@ -27,6 +27,10 @@
  */
 public class JsonSettingsLoader extends XContentSettingsLoader {
 
+    public JsonSettingsLoader(boolean guardAgainstNullValuedSettings) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/common/settings/loader/JsonSettingsLoader.java]: is there any place we want this to be false? just wondering if we just make it default behavior...

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/common/settings/loader/JsonSettingsLoader.java]: yeah, we do ... never mind :(

**Comment by bleskes** [comment]
nice. Happy we found a better name for the exception than the dreaded NPE

File: core/src/main/java/org/elasticsearch/common/settings/loader/SettingsLoaderFactory.java [hunk 2]
```diff
@@ -20,43 +20,63 @@
...
+     * fails this condition but contains a ':' then the content should
+     * be parsed as YAML, and otherwise should be parsed as properties.
+     * Note that the JSON and YAML parsers returned by this method will
+     * accept null-valued keys.
+     *
+     * @param source The underlying settings content.
+     * @return A settings loader.
      */
     public static SettingsLoader loaderFromSource(String source) {
         if (source.indexOf('{') != -1 && source.indexOf('}') != -1) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/settings/loader/SettingsLoaderFactory.java]: This detection feels pretty similar to `XContentFactory#xContentType(CharSequence)`. Maybe we can use that?

**Comment by jasontedor** [comment]
> LGTM. I would just call the boolean "allowNullValues" though.

I agree that that is better and I pushed 4d27328a8337ef1214aa7637c0247103b3835d40.

File: core/src/main/java/org/elasticsearch/common/settings/loader/SettingsLoaderFactory.java (hunk 2 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/settings/loader/SettingsLoaderFactory.java]: @nik9000 I have a code cleanup PR in the works related to this; I'll investigate that change in the context of that PR.

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/settings/loader/SettingsLoaderFactory.java]: I don't think that we should do this; the YAML detection there is different and I'm hesitant to change it.

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/settings/loader/SettingsLoaderFactory.java]: I'm fine with that. It'd help me the next time I look at this code if you add a comment explaining why you don't think this should share the detection logic. Or else I'm going to think the same thing all over again and waste time coming to the same conclusions you have. Or worse, I won't see why you didn't want to merge them and I'll break something.

**Pull request merged**